transaction/core/tests/input_rules.rs

```diff
@@ -63,25 +72,290 @@ fn test_input_rules_verify_max_tombstone() {
+            block_version,
+            4000,
+            INITIALIZE_LEDGER_AMOUNT - 4000,
+            &[&alice_pub, &alice_pub, &alice_pub, &alice_pub],
+        );
+
+    // Modify the Tx to have some (empty) input rules.
+    // (This invalidates the signature, but we aren't checking that here)
+    tx.prefix.inputs[0].input_rules = Some(InputRules::default());
+
+    // Check that the Tx is following input rules (vacuously)
+    get_first_rules(&tx).verify(block_version, &tx).unwrap();
+
+    // We're going to pull the four TxOut's from the TxPrefix, and make
+    // corresponding fractional outputs and fractional change from them, also
+    // rewriting their amounts when we do that, to see what happens to the
+    // validation routine. In order to do that, we need to get their amount
+    // shared secrets, and since Alice is the owner, we need Alice's
+    // account keys to do that.
+    assert_eq!(tx.prefix.outputs.len(), 4);
+    assert!(tx.prefix.outputs[0].public_key != tx.prefix.outputs[1].public_key);
+    let revealed_tx_outs: Vec<RevealedTxOut> = tx
+        .prefix
+        .outputs
+        .iter()
+        .map(|txo| {
+            let decompressed_tx_pub = RistrettoPublic::try_from(&txo.public_key).unwrap();
+            let tx_out_shared_secret =
+                get_tx_out_shared_secret(alice.view_private_key(), &decompressed_tx_pub);
+            let amount_shared_secret =
+                MaskedAmount::compute_amount_shared_secret(block_version, &tx_out_shared_secret)
+                    .unwrap();
+            let (amount, _) = txo
+                .masked_amount
+                .as_ref()
+                .unwrap()
+                .get_value_from_amount_shared_secret(&amount_shared_secret)
+                .unwrap();
+            assert_eq!(amount.value, 1000);
+            assert_eq!(*amount.token_id, 0);
+            RevealedTxOut {
+                tx_out: txo.clone(),
+                amount_shared_secret: amount_shared_secret.to_vec(),
+            }
+        })
+        .collect();
+
+    // Add a fractional output, by doubling one of the revealed tx outs.
+    // So this means we are filling the 1/2 of the request.
+    // Before we add a fractional change output, this is ill-formed, we can only
+    // verify fractional output rules if there is a fractional change output,
+    // because that is how we deduce the fill fraction.
+    get_first_rules_mut(&mut tx)
+        .fractional_outputs
+        .push(change_committed_amount(
+            &revealed_tx_outs[1],
+            Amount::new(2000, 0.into()),
+        ));
+    assert_matches!(
+        get_first_rules(&tx).verify(block_version, &tx),
+        Err(InputRuleError::FractionalOutputsNotExpected)
+    );
+
+    // Add a fractional input, also by doubling one of the revealed tx outs.
+    // This means the fill fraction is 1/2, which we are satisfying.
+    get_first_rules_mut(&mut tx).fractional_change = Some(change_committed_amount(
```

Comment:
I think the way this term arose is the following:

* We have well established concepts of "outputs" and "change outputs" (at least since RTH, these are the names of specific methods on `TransactionBuilder`)
* In MCIP 31 we added the concept of "required outputs", which appear in rules of SCIs. We naturally then had the concepts of "required outputs" and "required change outputs", which are the names of specific methods on `SignedContingentInputBuilder`. `add_required_output` is called when you want a required output with the Authenticated Sender memo, and `add_required_change_output` is called when you want a required output with the Destination Memo, which is destined for yourself.
* In the new MCIP we have a different type of constraint that can appear in InputRules, and that's this whole partial-fill concept. So naturally we now have `fractional_outputs` which appear in the rules, have amounts which vary according to the fill fraction, and have Authenticated Sender memo, and we have `fractional_change_outputs` which appear in the rules, have amounts which vary according to the fill fraction, and have Destination memo.

I'm happy to consider alternate names, but let's try to make the names all line up if we can.

It may be that "fractional" is just not the right adjective to use here. We could say that instead of "fractional output" and "fractional change" it's just "required partial fill output" and "required partial fill change". I would prefer something a little shorter.

Reply:
from the other comment:

> All of the "fractional" values are proposed or idealized values, and the final txouts are the partial or fractional values.
>
> Getting here I think I finally understand that txout_0 is Alice's "change" from the money she fronted

Yes -- anything that appears in "rules" is not real, only `TxPrefix.outputs` are real and actually enter the blockchain. The terms used in docs and the MCIP all speak of the "real output corresponding to the fractional output" which appears in the rules.

Looking back I currently believe this is well documented, but if you think there's a place I could have documented this better or you want to propose different naming, I'm open to it.